tasks/eva_2205/src/main/groovy/eva2205/eva2205_undo_merges_into_multimaps.groovy

```diff
@@ -0,0 +1,213 @@
+                    .and("inactiveObjects.hashedMessage").in(ssHashesWithMapWt)), it)}.flatten()
+        def impactedSSOpHistory = ssOpsInvolvedInMergeIntoMultimaps.groupBy { it.inactiveObjects[0].hashedMessage }
+        def recentSSRecordsWithoutMapWt = impactedSSOpHistory.collectEntries { ssHash, opsForSSHash ->
+            def mergeChain = getChronologicalMergeChain(opsForSSHash)
+            if (mergeChain.size() > 1) {
+                logger.info("These operations involve more than one merge in a merge chain:" + mergeChain)
+            }
+            def originalSSRecord = mergeChain[0]
+            // Return the most recent operation that recorded the SS without a map-weight,
+            // if the original SS did not have a map-weight
+            if (Objects.nonNull(originalSSRecord) && Objects.isNull(originalSSRecord.inactiveObjects[0].mapWeight)) {
+                return [ssHash, mostRecentNonMapWtSSRecord(mergeChain)]
+            }
+            return [ssHash, null]
+        }.findAll { k, v -> Objects.nonNull(v) }
+        undoMultiMapRSAssignment(recentSSRecordsWithoutMapWt, mapWtSVEs.groupBy{it.hashedMessage})
+        return recentSSRecordsWithoutMapWt.values().flatten()
+    }
+
+    def deprecateAndClearSSBatch(ssToDeprecate) {
+        DeprecateMapWtSS.deprecateSS(dbEnv, ssToDeprecate.toSet()) // Use unique to weed out duplicates within a batch
+        ssToDeprecate.clear()
+    }
+
+    def deprecateOrphanedCvesWithSameHash(cvesToResurrect) {
+        def cvesWithSameHash = [cveClass, dbsnpCveClass].collect{coll ->
+            dbEnv.mongoTemplate.find(query(where("_id").in(cvesToResurrect.collect{it.hashedMessage}).and(
+                    "mapWeight").exists(true)), coll)
+        }.flatten()
+        def cvDeprecationWriter = new ClusteredVariantDeprecationWriter(assembly,
+                dbEnv.mongoTemplate,
+                dbEnv.submittedVariantAccessioningService,
+                dbsnpRSIDUpperBound,
+                "EVA2205", "Variant deprecated due to mapWeight > 1")
+        // By the design of the write, this will only succeed if these RS IDs are orphaned i.e., there are no SS that have these RS IDs
+        cvDeprecationWriter.write(cvesWithSameHash)
+        this.metricCompute.addCount(ClusteringMetric.CLUSTERED_VARIANTS_DEPRECATED,
+                cvDeprecationWriter.numDeprecatedEntities)
+        this.metricCompute.saveMetricsCountsInDB()
+    }
+
+    // See https://docs.google.com/spreadsheets/d/1kRKHR4zrq-nxH_Sg82TwXZbxdgjB3K-q4YMyJva8yMg/edit#rangeid=1814908308
+    def undoMergesIntoMultiMapRS() {
+        def mapWtSSCursors = [sveClass, dbsnpSveClass].collect {
+            new RetryableBatchingCursor(where("seq").is(assembly).and("mapWeight").exists(true), dbEnv.mongoTemplate, it)
+        }
+        def collectionWithCvesToResurrect = "cvesToResurrect"
+        def svesStillWithMapWtRS = new ArrayList<>()
+        def batchIndex = 0
+        mapWtSSCursors.each {it.each {mapWtSVEs ->
```

Comment:
I've been trying to think through the scenario where multiple SS have been assigned an RS that have been merged into a mapWeight RS. 
Aren't we going to create multiple undoMerge event and attempt to resurrect the clustered variant multiple time ?

Reply:
Or is it the `bulkInsertIgnoreDuplicates` that will essentially ignore the entries that already exists.

Reply:
Yes, bulkInsertIgnoreDuplicates will ignore existing entries...